pkg/api/e2e_test.go

```diff
@@ -0,0 +1,93 @@
+//go:build integration
+// +build integration
+
+package api_test
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	"go.anx.io/go-anxcloud/pkg/api"
+	clouddnsv1 "go.anx.io/go-anxcloud/pkg/apis/clouddns/v1"
+	corev1 "go.anx.io/go-anxcloud/pkg/apis/core/v1"
+	vlanv1 "go.anx.io/go-anxcloud/pkg/apis/vlan/v1"
+	"go.anx.io/go-anxcloud/pkg/client"
+	"go.anx.io/go-anxcloud/pkg/utils/test"
+	testutils "go.anx.io/go-anxcloud/pkg/utils/test"
+)
+
+const (
+	waitTimeout  = 5 * time.Minute
+	retryTimeout = 15 * time.Second
+)
+
+var _ = Describe("options", func() {
+	var a api.API
+
+	BeforeEach(func() {
+		var err error
+		a, err = api.NewAPI(
+			api.WithClientOptions(client.AuthFromEnv(false)),
+		)
+		Expect(err).ToNot(HaveOccurred())
+	})
+
+	Context("AutoTag", func() {
+		It("can automatically tag resources on api.Create", func() {
+			vlan := vlanv1.VLAN{
+				DescriptionCustomer: "go-anxcloud test api.Create AutoTag " + testutils.RandomHostname(),
+				Locations: []corev1.Location{
+					{Identifier: "52b5f6b2fd3a4a7eaaedf1a7c019e9ea"},
+				},
+			}
+
+			DeferCleanup(func() {
+				if vlan.Identifier == "" {
+					return
+				}
+
+				Eventually(func(g Gomega) {
+					err := a.Get(context.TODO(), &vlan)
+					g.Expect(err).NotTo(HaveOccurred())
+					g.Expect(vlan.Status).To(Equal(vlanv1.StatusActive))
+				}, waitTimeout, retryTimeout).Should(Succeed())
+
+				err := a.Destroy(context.TODO(), &vlan)
+				Expect(err).ToNot(HaveOccurred())
+			})
+
+			err := a.Create(context.TODO(), &vlan, api.AutoTag("foo", "bar", "baz"))
+			Expect(err).ToNot(HaveOccurred())
+
+			tags, err := corev1.ListTags(context.TODO(), a, &vlan)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(tags).To(ContainElements("foo", "bar", "baz"))
+
+		})
+
+		It("returns an error when resource tagging failed", func() {
+			zone := clouddnsv1.Zone{
+				Name:       test.RandomHostname() + ".go-anxcloud.test",
+				IsMaster:   true,
+				DNSSecMode: "unvalidated",
+				AdminEmail: "admin@go-anxcloud.test",
+				Refresh:    14400,
+				Retry:      3600,
+				Expire:     604800,
+				TTL:        3600,
+			}
+
+			DeferCleanup(func() {
+				_ = a.Destroy(context.TODO(), &zone)
+			})
+
+			err := a.Create(context.TODO(), &zone, api.AutoTag("foo", "bar", "baz"))
+			Expect(err).To(HaveOccurred())
+			var e *api.ErrTaggingFailed
+			Expect(errors.As(err, &e)).To(BeTrue())
+		})
```

Comment:
this fails because we cannot tag CloudDNS zones, correct? We should add a comment here why we expect this to fail - just to make that clear for future readers :)

Reply:
thanks - another question: with our recently added custom identifiers feature, could we add support for the resource identifier to CloudDNS now?

Reply:
Not without using corev1.Resource because CloudDNS does not expose the common identifier at all.

Reply:
yeah, was thinking of adding an `Identifier` attribute to the Zone type, which is empty most of the time and have our custom identifier handling return the name for all operations - except for tagging it would take that identifier.

With that, we could at least manually set the identifier and tag a Zone and are prepared for Engine CDNS API returning the Identifier.

Reply:
We'd probably need to add an additional operation to make this work. Manual tagging of zones can be done by using `corev1.Tag(ctx, apiClient, &corev1.Resource{Identifier: "<common zone identifier>"})`.